src/app/shared/services/subsigners/metamask-subsigner.service.ts

```diff
@@ -13,45 +12,56 @@ export class MetamaskSubsignerService implements Subsigner {
   constructor(private preferenceStore: PreferenceStore) {
   }
 
-  login(opts: SubsignerLoginOpts): Observable<ethers.providers.JsonRpcSigner> {
+  login(opts: SubsignerLoginOpts): Observable<providers.JsonRpcSigner> {
+    return this.getSigner().pipe(
+      concatMap(signer => this.checkChainID(signer, opts)),
+      concatMap(signer => this.loginGetAddress(signer, opts).pipe(
+        tap(address => this.preferenceStore.update({address, authProvider: AuthProvider.METAMASK})),
+        map(() => signer)
+      ))
+    )
+  }
+
+  private getSigner(): Observable<providers.JsonRpcSigner> {
     return of((window as any)?.ethereum).pipe(
       concatMap(web3Provider => web3Provider ?
-        of(new ethers.providers.Web3Provider(web3Provider)
-          .getSigner()) : throwError('NO_METAMASK')),
-      concatMap(signer => from(signer.getChainId()).pipe(
-        concatMap(chainID => chainID === this.preferenceStore.getValue().chainID ?
-          of(chainID) : from(signer.provider.send('wallet_addEthereumChain',
+        of(new providers.Web3Provider(web3Provider)
+          .getSigner()) : throwError('NO_METAMASK'))
```

Comment:
Why mention only metamask, is metamask only provider/signer?

Reply:
I'm mentioning Metamask because I'm dealing with Metamask in this piece of code (`metamask-subsigner.service.ts`).

These errors will be refactored in the future. I throw just a string to see in the console what happened. For now, I'm not handling these error because there are just two scenarios for now:
1. Soft login. When the app starts, it tries to log in with the cached wallet/authProvider pair. If anything goes wrong, I'm logging out the user.
2. Force login. When the user manually triggers the login button, these errors could be thrown. As I said, I'll handle it in the future.